File: maheshwari_motors_app/lib/app/data/models/purchase_model.dart
import 'purchase_item_model.dart';

class PurchaseModel {
  final String id;
  final int? numericId;
  final String purchaseNo;
  final String? supplierId;
  final String? supplierName;
  final DateTime date;
  final List<PurchaseItemModel> items;
  final String purchaseType;
  final double amount;
  final String paymentStatus;
  final double paidAmount;

  PurchaseModel({
    required this.id,
    this.numericId,
    required this.purchaseNo,
    this.supplierId,
    this.supplierName,
    required this.date,
    this.items = const [],
    required this.purchaseType,
    required this.amount,
    this.paymentStatus = 'due',
    this.paidAmount = 0,
  });

  double get balanceAmount => amount - paidAmount;

  factory PurchaseModel.fromJson(Map<String, dynamic> json) {
    final supplier = json['supplier_id'];
    return PurchaseModel(
      id: json['_id'] ?? '',
      numericId: json['id'],
      purchaseNo: json['purchase_no'] ?? '',
      supplierId: supplier is Map ? supplier['_id'] : supplier?.toString(),
      supplierName: supplier is Map ? supplier['name'] : null,
      date: DateTime.tryParse(json['date'] ?? '') ?? DateTime.now(),
      items: json['items'] != null
          ? (json['items'] as List)
                .map((e) => PurchaseItemModel.fromJson(e))
                .toList()
          : [],
      purchaseType: json['purchase_type'] ?? 'GST',
      amount: (json['amount'] ?? 0).toDouble(),
      paymentStatus: json['payment_status'] ?? 'due',
      paidAmount: (json['paid_amount'] ?? 0).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{
      '_id': id,
      'purchase_no': purchaseNo,
      'date': date.toIso8601String(),
      'items': items.map((e) => e.toJson()).toList(),
      'purchase_type': purchaseType,
      'amount': amount,
      'payment_status': paymentStatus,
      'paid_amount': paidAmount,
    };
    if (supplierId != null) map['supplier_id'] = supplierId;
    return map;
  }
}

File: backend/src/models/purchase.model.js
import mongoose from "mongoose";

const purchaseSchema = new mongoose.Schema(
  {
    purchase_no: { type: String, required: true },
    date: { type: Date, required: true, default: Date.now },
    supplier_id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Supplier",
      required: true,
    },
    items: [
      {
        item_id: {
          type: mongoose.Schema.Types.ObjectId,
          ref: "Item",
          required: true,
        },
        quantity: { type: Number, required: true },
        rate: { type: Number, required: true },
        amount: { type: Number, required: true },
      },
    ],
    purchase_type: { type: String, enum: ["GST", "NON_GST"], required: true },
    amount: { type: Number, required: true },
    payment_status: {
      type: String,
      enum: ["due", "paid", "overpaid"],
      default: "due",
    },
    paid_amount: { type: Number, default: 0 },
    firm_id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Firm",
      required: true,
    },
    user_id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
    },
  },
  { timestamps: true },
);

purchaseSchema.index({ purchase_no: 1, firm_id: 1 }, { unique: true });

export default mongoose.model("Purchase", purchaseSchema);
